src/net_processing.cpp

```diff
@@ -1678,6 +1699,63 @@ void static ProcessGetBlockData(CNode& pfrom, const CChainParams& chainparams, c
     }
 }
 
+class CMWEBLeafsetMsg
+{
+public:
+    CMWEBLeafsetMsg() = default;
+    CMWEBLeafsetMsg(uint256 block_hash_in, BitSet leafset_in)
+        : block_hash(std::move(block_hash_in)), leafset(std::move(leafset_in)) { }
+
+    SERIALIZE_METHODS(CMWEBLeafsetMsg, obj) { READWRITE(obj.block_hash, obj.leafset); }
+
+    uint256 block_hash;
+    BitSet leafset;
+};
+
+static void ProcessGetMWEBLeafset(CNode& pfrom, const CChainParams& chainparams, const CInv& inv, CConnman& connman)
+{
+    ActivateBestChainIfNeeded(chainparams, inv);
+
+    LOCK(cs_main);
+    CBlockIndex* pindex = LookupBlockIndex(inv.hash);
+    if (!pindex) {
+        return;
+    }
+
+    // TODO: Add an outbound limit
+
+    // Avoid leaking prune-height by never sending blocks below the NODE_NETWORK_LIMITED threshold
+    if (::ChainActive().Tip()->nHeight - pindex->nHeight > MAX_MWEB_LEAFSET_DEPTH) {
+        LogPrint(BCLog::NET, "Ignore block request below MAX_MWEB_LEAFSET_DEPTH threshold from peer=%d\n", pfrom.GetId());
+
+        // disconnect node and prevent it from stalling (would otherwise wait for the MWEB leafset)
+        if (!pfrom.HasPermission(PF_NOBAN)) {
+            pfrom.fDisconnect = true;
+        }
+
+        return;
+    }
+
+    // Pruned nodes may have deleted the block, so check whether
+    // it's available before trying to send.
+    if (pindex->nStatus & BLOCK_HAVE_DATA && pindex->nStatus & BLOCK_HAVE_MWEB) {
+        // Rewind leafset to block height
+        BlockValidationState state;
+        CCoinsViewCache temp_view(&::ChainstateActive().CoinsTip());
+        if (!ActivateArbitraryChain(state, temp_view, chainparams, pindex)) {
+            pfrom.fDisconnect = true;
```

Comment:
Does this need to check PF_NOBAN like above?

Reply:
That's mainly just for indicating that the peer shouldn't be dropped for requesting too much data or otherwise misbehaving, but tbh, it's not used consistently at all throughout the code. The flows where I don't check the permission are the ones which shouldn't even be possible (it would require the node state to be corrupted), so I think it's probably fine that we don't check in those cases.